src/main/kotlin/org/rust/lang/core/ControlFlowGraph.kt

```diff
@@ -0,0 +1,199 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.lang.core
+
+import com.intellij.psi.PsiElement
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.ext.RsElement
+import org.rust.lang.core.psi.ext.ancestors
+import org.rust.lang.core.psi.ext.valueParameters
+import org.rust.lang.core.types.ty.TyNever
+import org.rust.lang.core.types.type
+
+sealed class CFGNodeData(val element: RsElement? = null) {
+    class AST(element: RsElement) : CFGNodeData(element)
+    object Entry : CFGNodeData()
+    object Exit : CFGNodeData()
+    object Dummy : CFGNodeData()
+    object Unreachable : CFGNodeData()
+
+    val text: String
+        get() = when (this) {
+            is AST -> element?.cfgText()?.trim() ?: "AST"
+            is Entry -> "Entry"
+            is Exit -> "Exit"
+            is Dummy -> "Dummy"
+            is Unreachable -> "Unreachable"
+        }
+
+    private fun RsElement.cfgText(): String =
+        when (this) {
+            is RsBlock, is RsBlockExpr -> "BLOCK"
+            is RsIfExpr -> "IF"
+            is RsWhileExpr -> "WHILE"
+            is RsLoopExpr -> "LOOP"
+            is RsForExpr -> "FOR"
+            is RsMatchExpr -> "MATCH"
+            is RsExprStmt -> expr.cfgText() + ";"
+            else -> this.text
+        }
+}
+
+class CFGEdgeData(val exitingScopes: List<RsElement>)
+
+typealias CFGGraph = Graph<CFGNodeData, CFGEdgeData>
+typealias CFGNode = Node<CFGNodeData, CFGEdgeData>
+typealias CFGEdge = Edge<CFGNodeData, CFGEdgeData>
```

Comment:
CFGGraph & CFGEdge are not used anywhere

Reply:
Got rid of `CFGGraph`. `CFGEdge` will be used in data-flow analysis.